sr_port/f_query.c

```diff
@@ -10,95 +13,254 @@
 
 #include "mdef.h"
+
 #include "compiler.h"
 #include "opcode.h"
 #include "indir_enum.h"
 #include "toktyp.h"
 #include "mdq.h"
+#include "mmemory.h"
 #include "fullbool.h"
+#include "fnorder.h"
+#include "advancewindow.h"
+#include "glvn_pool.h"
+
+GBLREF	boolean_t	run_time;
+GBLREF	int		source_column;
 
+error_def(ERR_QUERY2);
+error_def(ERR_SIDEEFFECTEVAL);
 error_def(ERR_VAREXPECTED);
 
 int f_query(oprtype *a, opctype op)
 {
-	triple		*oldchain, *r, *r0, *r1;
+	boolean_t	ok;
+	enum order_dir	direction;
+	enum order_obj	object;
+	int4		intval;
+	opctype		gv_oc;
+	oprtype		*tmpa, *tmpa0, *tmpa1, tmpa2;
+	oprtype		control_slot, dir_opr, *dir_oprptr, *next_oprptr;
 	save_se		save_state;
+	short int	column;
+	triple		*chain2, *obp, tmpchain2;
+	triple		*oldchain, *r, *r0, *r1, *r2, *sav_dirref, *triptr;
+	triple		*sav_gv1, *sav_gvn, *sav_lvn, *sav_ref;
+	opctype		query_opc[LAST_OBJECT][LAST_DIRECTION] =
+	{
+		/* FORWARD	BACKWARD		TBD */
+		{ OC_GVQUERY,	OC_GVREVERSEQUERY,	OC_GVQ2		},	/* GLOBAL */
+		{ OC_FNQUERY,	OC_FNREVERSEQUERY,	OC_FNQ2		},	/* LOCAL */
+		{ 0,		0,			0		},	/* LOCAL_NAME */
+		{ OC_INDFUN,	OC_INDFUN,		OC_INDQ2	}	/* INDIRECT */
+	};
 	DCL_THREADGBL_ACCESS;
 
 	SETUP_THREADGBL_ACCESS;
-	if (TK_IDENT == TREF(window_token))
+	/* Note: The below code is similar to f_query1.c. Any changes here might need to be made there and vice versa */
+	oldchain = sav_dirref = NULL;			/* default to no direction and no shifting indirection */
+	r = maketriple(OC_NOOP);			/* We'll fill in the opcode later, when we figure out what it is */
+	switch (TREF(window_token))
 	{
-		if (!lvn(a, OC_FNQUERY, 0))
-			return FALSE;
-		assert(TRIP_REF == a->oprclass);
-		if (OC_FNQUERY == a->oprval.tref->opcode)
+	case TK_IDENT:
+		object = LOCAL;
+		tmpa = &r->operand[0];
+		ok = lvn(tmpa, OC_FNQUERY, 0);
+		if (!ok)
+			break;
+		assert(TRIP_REF == tmpa->oprclass);
+		tmpa0 = &tmpa->oprval.tref->operand[0];
+		if (OC_FNQUERY == tmpa->oprval.tref->opcode)
 		{
-			assert(OC_FNQUERY == a->oprval.tref->opcode);
-			assert(TRIP_REF == a->oprval.tref->operand[0].oprclass);
-			assert(OC_ILIT == a->oprval.tref->operand[0].oprval.tref->opcode);
-			assert(ILIT_REF == a->oprval.tref->operand[0].oprval.tref->operand[0].oprclass);
-			assert(0 < a->oprval.tref->operand[0].oprval.tref->operand[0].oprval.ilit);
-			a->oprval.tref->operand[0].oprval.tref->operand[0].oprval.ilit += 2;
-			assert(TRIP_REF == a->oprval.tref->operand[1].oprclass);
-			assert(OC_PARAMETER == a->oprval.tref->operand[1].oprval.tref->opcode);
-			assert(TRIP_REF == a->oprval.tref->operand[1].oprval.tref->operand[0].oprclass);
-			r0 = a->oprval.tref->operand[1].oprval.tref->operand[0].oprval.tref;
+			assert(TRIP_REF == tmpa0->oprclass);
+			assert(OC_ILIT == tmpa0->oprval.tref->opcode);
+			assert(ILIT_REF == tmpa0->oprval.tref->operand[0].oprclass);
+			assert(0 < tmpa0->oprval.tref->operand[0].oprval.ilit);
+			tmpa0->oprval.tref->operand[0].oprval.ilit += 2;
```

Comment:
Macro seems overkill since this usage is specific (only one place). Added comment. Note that this and the previous comment were for code that previously existed.